File: algorithm/leetcode5438.minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    int minDays(vector<int>& bloomDay, int m, int k) {
        if(m*k > bloomDay.size()) return -1;
        int l = 1, r = 1e9, mid;
        while(l < r)
        {
            mid = l+((r-l)>>1);
            if(haveflower(bloomDay, mid, m, k))
                r = mid;
            else
                l = mid+1;
        }
        return l;
    }
    bool haveflower(vector<int>& day, int days, int m, int k) 
    {
        int flowerPack = 0, flowercount = 0;
        for(auto d : day)
        {
            if(days/d >= 1)//能开花
                flowercount++;
            else
                flowercount = 0;
            if(flowercount == k)
            {
                flowerPack++;
                flowercount = 0;
            }
        }
        return flowerPack >= m;
    }
};

File: algorithm/Python3_version/leetcode5438.minimum-number-of-days-to-make-m-bouquets.py
class Solution:
    def minDays(self, bloomDay: List[int], m: int, k: int) -> int:
        if m*k > len(bloomDay):
            return -1
        def haveflower(days):
            flowerPack, flowercount = 0, 0
            for d in bloomDay:
                if days//d >= 1:
                    flowercount += 1
                else:
                    flowercount = 0
                if flowercount == k:
                    flowerPack += 1
                    flowercount = 0
            return flowerPack >= m
        l, r = 1, int(1e9)
        while l < r:
            mid = l+((r-l)>>1)
            if(haveflower(mid)):
                r = mid
            else:
                l = mid+1
        return l